File: ThreeSum.js
var threeSum = function(nums) {
    nums.sort((a, b) => a - b);
    const result = [];
    for (var i = 0; i < nums.length; i++) {
        if (i > 0 && nums[i] === nums[i - 1]) {
            continue;
        }

        var left = i + 1;
        var right = nums.length - 1;

        while (left < right) {
            const threeSum = nums[i] + nums[left] + nums[right];
            if (threeSum > 0) {
                right--;
            } else if (threeSum < 0) {
                left++;
            } else {
                result.push([nums[i], nums[left], nums[right]]);
                left++;
                while (nums[left] === nums[left - 1] && left < right) {
                    left++;
                }
            }
        }
    } 
    return result;
};

File: ThreeSum.py
class ThreeSum(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums.sort()
        if nums[0] > 0:
            return []
        
        result = []
        for i in range (len(nums)):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            
            left = i + 1
            right = len(nums) - 1
            
            while left < right:
                threeSum = nums[i] + nums[left] + nums[right]
                if threeSum > 0:
                    right-=1
                elif threeSum < 0:
                    left+=1
                else:
                    triplet = [nums[i], nums[left], nums[right]]
                    result.append(triplet)
                    left+=1
                    while nums[left] == nums[left - 1] and left < right:
                        left+=1
                        
        return result